Automated Smoketest: :x:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 Host: ip-10-20-0-74.ca-central-1.compute.internal
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Git: 2.9.5,  CMake: 3.22.1,  cUrl: 7.67.0,  node.js: v16.13.1,  npm: 8.1.2
Sha: ba2905bc21994df7ae32203e088f82944017fc36
Containerized:False
Number of Error(s): 5
src/ESPSearch.ts(37,25): error TS6138: Property begin is declared but its value is never read.
src/ESPSearch.ts(37,99): error TS6138: Property end is declared but its value is never read.
make[2]: *** [esp/src/CMakeFiles/eclwatch.dir/build.make:330: esp/src/build/dist/stub.eclwatch.js] Error 1
make[1]: *** [CMakeFiles/Makefile2:12228: esp/src/CMakeFiles/eclwatch.dir/all] Error 2
make: *** [Makefile:166: all] Error 2
Get ECLWatch build result
ReportTimes.
Finish at: 2022-02-23_22-14-26
time logs started
Prep time   :    23 sec (00:00:23)
Build time  :   158 sec (00:02:38)
Package time:     0 sec (00:00:00)
Install time:     0 sec (00:00:00)
Start time  :     0 sec (00:00:00)
Test time   :     0 sec (00:00:00)
Stop time   :     0 sec (00:00:00)
Summary     :   181 sec (00:03:01)
time logs end

[Build error:](https://gist.github.com/83f3e67f3ed88dc63c339f21936a4506)
Build log [Build log](https://gist.github.com/HPCCSmoketest/83f3e67f3ed88dc63c339f21936a4506#file-result-22-02-23-22-11-08-log)

--- a/esp/src/eclwatch/SearchResultsWidget.js
+++ b/esp/src/eclwatch/SearchResultsWidget.js
@@ -271,102 +271,138 @@ define([
             return null;
         },
 
-        loadWUQueryResponse: function (prefix, response) {
-            var workunits = lang.getObject("WUQueryResponse.Workunits.ECLWorkunit", false, response);
-            if (workunits) {
-                var idPrefix = prefix.split(" ").join("_");
-                var context = this;
-                arrayUtil.forEach(workunits, function (item, idx) {
-                    context.store.add({
-                        storeID: ++context._rowID,
-                        id: context.id + item.Wuid,
-                        Type: context.i18n.ECLWorkunit,
-                        Reason: prefix,
-                        Summary: item.Wuid,
-                        _type: "Wuid",
-                        _wuid: item.Wuid
-                    });
-                    context.eclStore.add(ESPWorkunit.Get(item.Wuid, item), { overwrite: true });
-                });
-                this.refreshTab(this.eclTab);
-                return workunits.length;
-            }
-            return 0;
+        parseWuQueryResponse: function (prefix, response) {
+            return response?.WUQueryResponse?.Workunits?.ECLWorkunit.map(item => {
+                return {
+                    storeID: ++this._rowID,
+                    id: item.Wuid,
+                    Type: nlsHPCC.ECLWorkunit,
+                    Reason: prefix,
+                    Summary: item.Wuid,
+                    _type: "Wuid",
+                    _wuid: item.Wuid,
+                    ...item
+                };
+            });
         },
 
-        loadGetDFUWorkunitsResponse: function (prefix, response) {
-            var workunits = lang.getObject("GetDFUWorkunitsResponse.results.DFUWorkunit", false, response);
-            if (workunits) {
-                var idPrefix = prefix.split(" ").join("_");
-                var context = this;
-                arrayUtil.forEach(workunits, function (item, idx) {
-                    context.store.add({
-                        storeID: ++context._rowID,
-                        id: context.id + item.ID,
-                        Type: context.i18n.DFUWorkunit,
+        parseGetDFUWorkunitResponse: function (prefix, response) {
+            const item = response?.GetDFUWorkunitResponse?.result;
+            return (item && item.State !== 999) ? {
+                storeID: ++this._rowID,
+                id: item.ID,
+                Type: nlsHPCC.DFUWorkunit,
+                Reason: prefix,
+                Summary: item.ID,
+                _type: "DFUWuid",
+                _wuid: item.ID,
+                ...item
+            } : undefined;
+        },
+
+        parseGetDFUWorkunitsResponse: function (prefix, response) {
+            return response?.GetDFUWorkunitsResponse?.results?.DFUWorkunit.map(item => {
+                return {
+                    storeID: ++this._rowID,
+                    id: item.ID,
+                    Type: nlsHPCC.DFUWorkunit,
+                    Reason: prefix,
+                    Summary: item.ID,
+                    _type: "DFUWuid",
+                    _wuid: item.ID,
+                    ...item
+                };
+            });
+        },
+
+        parseDFUQueryResponse: function (prefix, response) {
+            return response?.DFUQueryResponse?.DFULogicalFiles?.DFULogicalFile.map(item => {
+                if (item.isSuperfile) {
+                    return {
+                        storeID: ++this._rowID,
+                        id: item.Name,
+                        Type: nlsHPCC.SuperFile,
                         Reason: prefix,
-                        Summary: item.ID,
-                        _type: "DFUWuid",
-                        _wuid: item.ID
-                    });
-                    context.dfuStore.add(ESPDFUWorkunit.Get(item.ID, item), { overwrite: true });
+                        Summary: item.Name,
+                        _type: "SuperFile",
+                        _nodeGroup: item.NodeGroup,
+                        _name: item.Name,
+                        ...item
+                    };
+                }
+                return {
+                    storeID: ++this._rowID,
+                    id: item.NodeGroup + "::" + item.Name,
+                    Type: nlsHPCC.LogicalFile,
+                    Reason: prefix,
+                    Summary: item.Name + " (" + item.NodeGroup + ")",
+                    _type: "LogicalFile",
+                    _nodeGroup: item.NodeGroup,
+                    _name: item.Name,
+                    ...item
+                };
+            });
+        },
+
+        parseWUListQueriesResponse: function (prefix, response) {
+            return response?.WUListQueriesResponse?.QuerysetQueries?.QuerySetQuery.map(item => {
+                return {
+                    storeID: ++this._rowID,
+                    id: item.QuerySetId + "::" + item.Id,
+                    Type: nlsHPCC.Query,
+                    Reason: prefix,
+                    Summary: item.Name + " (" + item.QuerySetId + " - " + item.Id + ")",
+                    _type: "Query",
+                    _querySetId: item.QuerySetId,
+                    _id: item.Id,
+                    ...item
+                };
+            });
+        },
+
+        loadWUQueryResponse: function (prefix, response) {
+            const results = this.parseWuQueryResponse(prefix, response);
+            if (results) {
+                results.forEach((item, idx) => {
+                    this.store.put(item, { overwrite: true });
+                    this.eclStore.put(ESPWorkunit.Get(item._wuid, item), { overwrite: true });
                 });
-                this.refreshTab(this.dfuTab);
-                return workunits.length;
+                this.refreshTab(this.eclTab);
+                return results.length;
             }
             return 0;
         },
 
         loadGetDFUWorkunitResponse: function (prefix, response) {
-            var context = this;
-            var workunit = lang.getObject("GetDFUWorkunitResponse.result", false, response);
-            if (workunit && workunit.State !== 999) {
-                var idPrefix = prefix.split(" ").join("_");
-                this.store.add({
-                    storeID: ++context._rowID,
-                    id: context.id + workunit.ID,
-                    Type: context.i18n.DFUWorkunit,
-                    Reason: prefix,
-                    Summary: workunit.ID,
-                    _type: "DFUWuid",
-                    _wuid: workunit.ID
-                });
-                this.dfuStore.add(ESPDFUWorkunit.Get(workunit.ID, workunit), { overwrite: true });
+            const workunit = this.parseGetDFUWorkunitResponse(prefix, response);
+            if (workunit) {
+                this.store.put(workunit, { overwrite: true });
+                this.dfuStore.put(ESPDFUWorkunit.Get(workunit._wuid, workunit), { overwrite: true });
                 this.refreshTab(this.dfuTab);
                 return 1;
             }
             return 0;
         },
 
+        loadGetDFUWorkunitsResponse: function (prefix, response) {
+            const results = this.parseGetDFUWorkunitsResponse(prefix, response);
+            if (results) {
+                results.forEach((item, idx) => {
+                    this.store.put(item, { overwrite: true });
+                    this.dfuStore.put(ESPDFUWorkunit.Get(item._wuid, item), { overwrite: true });
+                });
+                this.refreshTab(this.dfuTab);
+                return results.length;
+            }
+            return 0;
+        },
+
         loadDFUQueryResponse: function (prefix, response) {
-            var items = lang.getObject("DFUQueryResponse.DFULogicalFiles.DFULogicalFile", false, response);
+            const items = this.parseDFUQueryResponse(prefix, response);
             if (items) {
-                var idPrefix = prefix.split(" ").join("_");
-                var context = this;
-                arrayUtil.forEach(items, function (item, idx) {
-                    if (item.isSuperfile) {
-                        context.store.add({
-                            storeID: ++context._rowID,
-                            id: context.id + item.Name,
-                            Type: context.i18n.SuperFile,
-                            Reason: prefix,
-                            Summary: item.Name,
-                            _type: "SuperFile",
-                            _name: item.Name
-                        });
-                    } else {
-                        context.store.add({
-                            storeID: ++context._rowID,
-                            id: context.id + item.Name,
-                            Type: context.i18n.LogicalFile,
-                            Reason: prefix,
-                            Summary: item.Name + " (" + item.NodeGroup + ")",
-                            _type: "LogicalFile",
-                            _nodeGroup: item.NodeGroup,
-                            _name: item.Name
-                        });
-                    }
-                    context.fileStore.add(ESPLogicalFile.Get(item.NodeGroup, item.Name, item), { overwrite: true });
+                items.forEach((item, idx) => {
+                    this.store.put(item, { overwrite: true });
+                    this.fileStore.put(ESPLogicalFile.Get(item._nodeGroup, item._name, item), { overwrite: true });
                 });
                 this.refreshTab(this.fileTab);
                 return items.length;
@@ -375,22 +411,11 @@ define([
         },
 
         loadWUListQueriesResponse: function (prefix, response) {
-            var items = lang.getObject("WUListQueriesResponse.QuerysetQueries.QuerySetQuery", false, response);
+            const items = this.parseWUListQueriesResponse(prefix, response);
             if (items) {
-                var idPrefix = prefix.split(" ").join("_");
-                var context = this;
-                arrayUtil.forEach(items, function (item, idx) {
-                    context.store.add({
-                        storeID: ++context._rowID,
-                        id: context.id + item.QuerySetId + "::" + item.Id,
-                        Type: context.i18n.Query,
-                        Reason: prefix,
-                        Summary: item.Name + " (" + item.QuerySetId + " - " + item.Id + ")",
-                        _type: "Query",
-                        _querySetId: item.QuerySetId,
-                        _id: item.Id
-                    });
-                    context.queryStore.add(ESPQuery.Get(item.QuerySetId, item.Id, item), { overwrite: true });
+                items.forEach((item, idx) => {
+                    this.store.put(item, { overwrite: true });
+                    this.queryStore.put(ESPQuery.Get(item._querySetId, item._id, item), { overwrite: true });
                 });
                 this.refreshTab(this.queryTab);
                 return items.length;
@@ -422,14 +447,16 @@ define([
                 this.selectChild(tab);
             }
 
-            const specificSearch = ESPSearch.searchAll(this.searchText,
-                (what, response) => { this.loadWUQueryResponse(what, response); },
-                (what, response) => { this.loadGetDFUWorkunitResponse(what, response); },
-                (what, response) => { this.loadGetDFUWorkunitsResponse(what, response); },
-                (what, response) => { this.loadDFUQueryResponse(what, response); },
-                (what, response) => { this.loadWUListQueriesResponse(what, response); },
+            const specificSearch = ESPSearch.searchAll(
+                this.searchText,
+                (prefix, response) => { this.loadWUQueryResponse(prefix, response); },
+                (prefix, response) => { this.loadGetDFUWorkunitResponse(prefix, response); },
+                (prefix, response) => { this.loadGetDFUWorkunitsResponse(prefix, response); },
+                (prefix, response) => { this.loadDFUQueryResponse(prefix, response); },
+                (prefix, response) => { this.loadWUListQueriesResponse(prefix, response); },
                 (searchCount) => { },
-                (success) => { this.standby.hide(); });
+                (success) => { this.standby.hide(); }
+            );
 
             switch (specificSearch) {
                 case "ecl":

--- a/esp/src/src-react/components/Search.tsx
+++ b/esp/src/src-react/components/Search.tsx
@@ -1,9 +1,9 @@
 import * as React from "react";
-import { CommandBar, ContextualMenuItemType, ICommandBarItemProps, Pivot, PivotItem, ProgressIndicator } from "@fluentui/react";
+import { CommandBar, ContextualMenuItemType, ICommandBarItemProps, Link, Pivot, PivotItem, ProgressIndicator } from "@fluentui/react";
 import { useConst } from "@fluentui/react-hooks";
 import { ESPSearch } from "src/ESPSearch";
 import nlsHPCC from "src/nlsHPCC";
-import { useGrid } from "../hooks/grid";
+import { useFluentGrid } from "../hooks/grid";
 import { HolyGrail } from "../layouts/HolyGrail";
 import { ShortVerticalDivider } from "./Common";
 import { selector } from "./DojoGrid";
@@ -51,54 +51,63 @@ export const Search: React.FunctionComponent<SearchProps> = ({
     const [uiState, setUIState] = React.useState({ ...defaultUIState });
     const [searchCount, setSearchCount] = React.useState(0);
 
+    const [data, setData] = React.useState<any[]>([]);
+
     //  Search
-    const search = useConst(new ESPSearch(
-        searchCount => {
-            setSearchCount(searchCount);
-        },
-        () => {
-            progress.value++;
-            refreshTable();
-        }, () => {
-            setSearchCount(0);
-        }));
+    const search = useConst(new ESPSearch(() => { progress.value++; }));
 
     //  Grid ---
-    const [Grid, selection, refreshTable, copyButtons] = useGrid({
-        store: search.store,
+    const [Grid, selection, copyButtons] = useFluentGrid({
+        data,
+        primaryID: "__hpcc_id",
+        sort: [{ attribute: "__hpcc_id" }],
         filename: "search",
         columns: {
             col1: selector({ width: 27, selectorType: "checkbox" }),
-            Type: {
+            What: {
                 label: nlsHPCC.What, width: 108, sortable: true,
-                formatter: function (type, row) {
-                    return "<a href='" + searchResultUrl(row) + "'>" + type + "</a>";
+                formatter: function (_, row) {
+                    return <Link href={`${searchResultUrl(row)}`}>{row.Type}</Link>;
                 }
             },
             Reason: { label: nlsHPCC.Where, width: 108, sortable: true },
             Summary: {
                 label: nlsHPCC.Who, sortable: true,
-                formatter: function (summary, row) {
-                    return "<a href='" + searchResultUrl(row) + "'>" + summary + "</a>";
+                formatter: function (Summary, row) {
+                    return <Link href={`${searchResultUrl(row)}`}>{Summary}</Link>;
                 }
             }
         }
     });
 
+    const refreshData = React.useCallback(() => {
+        search.searchAll(searchText).then(results => {
+            const _data = [];
+            const matchedIds: { [key: string]: boolean } = {};
+            results.forEach(resultCategory => {
+                if (resultCategory) {
+                    resultCategory.forEach(result => {
+                        if (!matchedIds[result.id]) {
+                            _data.push(result);
+                            matchedIds[result.id] = true;
+                        }
+                    });
+                }
+            });
+            setSearchCount(_data.length);
+            setData(_data);
+        });
+    }, [search, searchText]);
+
     React.useEffect(() => {
-        if (searchText) {
-            progress.value = 0;
-            setSearchCount(0);
-            search.searchAll(searchText);
-            refreshTable();
-        }
-    }, [progress, refreshTable, search, searchText]);
+        refreshData();
+    }, [refreshData]);
 
     //  Command Bar  ---
     const buttons = React.useMemo((): ICommandBarItemProps[] => [
         {
             key: "refresh", text: nlsHPCC.Refresh, iconProps: { iconName: "Refresh" },
-            onClick: () => refreshTable()
+            onClick: () => refreshData()
         },
         { key: "divider_1", itemType: ContextualMenuItemType.Divider, onRender: () => <ShortVerticalDivider /> },
         {
@@ -120,7 +129,7 @@ export const Search: React.FunctionComponent<SearchProps> = ({
                 setMine(!mine);
             }
         },
-    ], [mine, refreshTable, selection, uiState.hasSelection]);
+    ], [mine, refreshData, selection, uiState.hasSelection]);
 
     //  Selection  ---
     React.useEffect(() => {
@@ -137,7 +146,7 @@ export const Search: React.FunctionComponent<SearchProps> = ({
     return <HolyGrail
         header={<Pivot headersOnly={true} onLinkClick={(item: PivotItem) => setSelectedKey(item.props.itemKey!)
         }>
-            <PivotItem itemKey="all" headerText={nlsHPCC.All} itemCount={search.store.data.length} />
+            <PivotItem itemKey="all" headerText={nlsHPCC.All} itemCount={data.length} />
             <PivotItem itemKey="ecl" headerText={nlsHPCC.ECLWorkunit} headerButtonProps={search.eclStore.data.length === 0 ? disabled : undefined} itemCount={search.eclStore.data.length} />
             <PivotItem itemKey="dfu" headerText={nlsHPCC.DFUWorkunit} headerButtonProps={search.dfuStore.data.length === 0 ? disabled : undefined} itemCount={search.dfuStore.data.length} />
             <PivotItem itemKey="file" headerText={nlsHPCC.LogicalFile} headerButtonProps={search.fileStore.data.length === 0 ? disabled : undefined} itemCount={search.fileStore.data.length} />

--- a/esp/src/src/ESPSearch.ts
+++ b/esp/src/src/ESPSearch.ts
@@ -11,6 +11,16 @@ import nlsHPCC from "./nlsHPCC";
 
 export type searchAllResponse = undefined | "ecl" | "dfu" | "file" | "query";
 
+interface SearchParams {
+    searchECL: boolean;
+    searchECLText: boolean;
+    searchDFU: boolean;
+    searchFile: boolean;
+    searchQuery: boolean;
+    text: string;
+    searchType?: string;
+}
+
 export class ESPSearch {
 
     protected _searchID = 0;
@@ -24,12 +34,44 @@ export class ESPSearch {
     fileStore = Observable(new Memory("__hpcc_id"));
     queryStore = Observable(new Memory("__hpcc_id"));
 
-    constructor(private begin: (searchCount: number) => void, private update: () => void, private end: (success: boolean) => void) {
+    constructor(private update: () => void) {
     }
 
-    searchAll(searchText: string): searchAllResponse {
-        const searchID = ++this._searchID;
-        ++this.id;
+    generateSearchParams(searchText: string): SearchParams {
+        const searchParams: SearchParams = {
+            searchECL: false,
+            searchECLText: false,
+            searchDFU: false,
+            searchFile: false,
+            searchQuery: false,
+            text: searchText
+        };
+
+        if (searchText.indexOf("ecl:") === 0) {
+            searchParams.searchECL = true;
+            searchParams.searchECLText = true;
+            searchParams.text = searchText.substring(4);
+        } else if (searchText.indexOf("dfu:") === 0) {
+            searchParams.searchDFU = true;
+            searchParams.text = searchText.substring(4);
+        } else if (searchText.indexOf("file:") === 0) {
+            searchParams.searchFile = true;
+            searchParams.text = searchText.substring(5);
+        } else if (searchText.indexOf("query:") === 0) {
+            searchParams.searchQuery = true;
+            searchParams.text = searchText.substring(6);
+        } else {
+            searchParams.searchECL = true;
+            searchParams.searchDFU = true;
+            searchParams.searchFile = true;
+            searchParams.searchQuery = true;
+        }
+        searchParams.text = searchParams.text.trim();
+
+        return searchParams;
+    }
+
+    searchAll(searchText: string): any {
 
         this.store.setData([]);
         this.eclStore.setData([]);
@@ -37,234 +79,216 @@ export class ESPSearch {
         this.fileStore.setData([]);
         this.queryStore.setData([]);
 
-        let searchECL = false;
-        let searchECLText = false;
-        let searchDFU = false;
-        let searchFile = false;
-        let searchQuery = false;
-
-        let retVal: searchAllResponse = undefined;
-
-        if (searchText.indexOf("ecl:") === 0) {
-            retVal = "ecl";
-            searchECL = true;
-            searchECLText = true;
-            searchText = searchText.substring(4);
-        } else if (searchText.indexOf("dfu:") === 0) {
-            searchDFU = true;
-            searchText = searchText.substring(4);
-        } else if (searchText.indexOf("file:") === 0) {
-            searchFile = true;
-            searchText = searchText.substring(5);
-        } else if (searchText.indexOf("query:") === 0) {
-            searchQuery = true;
-            searchText = searchText.substring(6);
-        } else {
-            searchECL = true;
-            searchDFU = true;
-            searchFile = true;
-            searchQuery = true;
-        }
-        searchText = searchText.trim();
-
         const searchArray = [];
-        if (searchECL) {
-            searchArray.push(WsWorkunits.WUQuery({ request: { Wuid: "*" + searchText + "*" }, suppressExceptionToaster: true }).then(response => {
-                if (searchID === this._searchID) {
-                    this.loadWUQueryResponse(nlsHPCC.WUID, response);
-                    this.update();
-                }
+        const searchParams = this.generateSearchParams(searchText);
+
+        if (searchParams.searchECL) {
+            searchArray.push(WsWorkunits.WUQuery({ request: { Wuid: "*" + searchParams.text + "*" } }).then(response => {
+                const results = this.parseWuQueryResponse(nlsHPCC.WUID, response);
+                this.loadWUQueryResponse(results);
+                return results;
             }));
             searchArray.push(WsWorkunits.WUQuery({ request: { Jobname: "*" + searchText + "*" } }).then(response => {
-                if (searchID === this._searchID) {
-                    this.loadWUQueryResponse(nlsHPCC.JobName, response);
-                    this.update();
-                }
+                const results = this.parseWuQueryResponse(nlsHPCC.JobName, response);
+                this.loadWUQueryResponse(results);
+                return results;
             }));
             searchArray.push(WsWorkunits.WUQuery({ request: { Owner: searchText } }).then(response => {
-                if (searchID === this._searchID) {
-                    this.loadWUQueryResponse(nlsHPCC.Owner, response);
-                    this.update();
-                }
+                const results = this.parseWuQueryResponse(nlsHPCC.Owner, response);
+                this.loadWUQueryResponse(results);
+                return results;
             }));
         }
-        if (searchECLText) {
-            searchArray.push(WsWorkunits.WUQuery({ request: { ECL: searchText } }).then(response => {
-                if (searchID === this._searchID) {
-                    this.loadWUQueryResponse(nlsHPCC.ECL, response);
-                    this.update();
-                }
+        if (searchParams.searchECLText) {
+            searchArray.push(WsWorkunits.WUQuery({ request: { ECL: searchParams.text } }).then(response => {
+                const results = this.parseWuQueryResponse(nlsHPCC.ECL, response);
+                this.loadWUQueryResponse(results);
+                return results;
             }));
         }
-        if (searchDFU) {
-            searchArray.push(FileSpray.GetDFUWorkunit({ request: { wuid: "*" + searchText + "*" }, suppressExceptionToaster: true }).then(response => {
-                if (searchID === this._searchID) {
-                    this.loadGetDFUWorkunitResponse(nlsHPCC.WUID, response);
-                    this.update();
-                }
+        if (searchParams.searchDFU) {
+            searchArray.push(FileSpray.GetDFUWorkunit({ request: { wuid: "*" + searchText + "*" } }).then(response => {
+                const results = this.parseGetDFUWorkunitResponse(nlsHPCC.ECL, response);
+                this.loadGetDFUWorkunitResponse(results);
+                return results;
             }));
             searchArray.push(FileSpray.GetDFUWorkunits({ request: { Jobname: "*" + searchText + "*" } }).then(response => {
-                if (searchID === this._searchID) {
-                    this.loadGetDFUWorkunitsResponse(nlsHPCC.JobName, response);
-                    this.update();
-                }
+                const results = this.parseGetDFUWorkunitsResponse(nlsHPCC.JobName, response);
+                this.loadGetDFUWorkunitsResponse(results);
+                return results;
             }));
             searchArray.push(FileSpray.GetDFUWorkunits({ request: { Owner: searchText } }).then(response => {
-                if (searchID === this._searchID) {
-                    this.loadGetDFUWorkunitsResponse(nlsHPCC.Owner, response);
-                    this.update();
-                }
+                const results = this.parseGetDFUWorkunitsResponse(nlsHPCC.Owner, response);
+                this.loadGetDFUWorkunitsResponse(results);
+                return results;
             }));
         }
-        if (searchFile) {
-            searchArray.push(WsDfu.DFUQuery({ request: { LogicalName: "*" + searchText + "*" } }).then(response => {
-                if (searchID === this._searchID) {
-                    this.loadDFUQueryResponse(nlsHPCC.LogicalName, response);
-                    this.update();
-                }
+        if (searchParams.searchFile) {
+            searchArray.push(WsDfu.DFUQuery({ request: { LogicalName: "*" + searchParams.text + "*" } }).then(response => {
+                const results = this.parseDFUQueryResponse(nlsHPCC.LogicalName, response);
+                this.loadDFUQueryResponse(results);
+                return results;
             }));
-            searchArray.push(WsDfu.DFUQuery({ request: { Owner: searchText } }).then(response => {
-                if (searchID === this._searchID) {
-                    this.loadDFUQueryResponse(nlsHPCC.Owner, response);
-                    this.update();
-                }
+            searchArray.push(WsDfu.DFUQuery({ request: { Owner: searchParams.text } }).then(response => {
+                const results = this.parseDFUQueryResponse(nlsHPCC.Owner, response);
+                this.loadDFUQueryResponse(results);
+                return results;
             }));
         }
-        if (searchQuery) {
-            searchArray.push(WsWorkunits.WUListQueries({ request: { QueryID: "*" + searchText + "*" } }).then(response => {
-                if (searchID === this._searchID) {
-                    this.loadWUListQueriesResponse(nlsHPCC.ID, response);
-                    this.update();
-                }
+        if (searchParams.searchQuery) {
+            searchArray.push(WsWorkunits.WUListQueries({ request: { QueryID: "*" + searchParams.text + "*" } }).then(response => {
+                const results = this.parseWUListQueriesResponse(nlsHPCC.ID, response);
+                this.loadWUListQueriesResponse(results);
+                return results;
             }));
-            searchArray.push(WsWorkunits.WUListQueries({ request: { QueryName: "*" + searchText + "*" } }).then(response => {
-                if (searchID === this._searchID) {
-                    this.loadWUListQueriesResponse(nlsHPCC.Name, response);
-                    this.update();
-                }
+            searchArray.push(WsWorkunits.WUListQueries({ request: { QueryName: "*" + searchParams.text + "*" } }).then(response => {
+                const results = this.parseWUListQueriesResponse(nlsHPCC.Name, response);
+                this.loadWUListQueriesResponse(results);
+                return results;
             }));
         }
 
-        //  Always true  ---
-        this.begin(searchArray.length);
-        Promise.all(searchArray).then(results => {
-            if (searchID === this._searchID)
-                this.end(true);
-        }).catch(e => {
-            if (searchID === this._searchID)
-                this.end(false);
+        this.update();
 
-        });
-
-        return retVal;
+        return Promise.all(searchArray);
     }
 
-    loadWUQueryResponse(prefix, response) {
-        const workunits = response?.WUQueryResponse?.Workunits?.ECLWorkunit;
-        if (workunits) {
-            workunits.forEach((item, idx) => {
-                this.store.add({
+    parseWuQueryResponse(prefix, response) {
+        return response?.WUQueryResponse?.Workunits?.ECLWorkunit.map(item => {
+            return {
+                storeID: ++this._rowID,
+                id: item.Wuid,
+                Type: nlsHPCC.ECLWorkunit,
+                Reason: prefix,
+                Summary: item.Wuid,
+                _type: "Wuid",
+                _wuid: item.Wuid,
+                ...item
+            };
+        });
+    }
+
+    parseGetDFUWorkunitResponse(prefix, response) {
+        const item = response?.GetDFUWorkunitResponse?.result;
+        return (item && item.State !== 999) ? {
+            storeID: ++this._rowID,
+            id: item.ID,
+            Type: nlsHPCC.DFUWorkunit,
+            Reason: prefix,
+            Summary: item.ID,
+            _type: "DFUWuid",
+            _wuid: item.ID,
+            ...item
+        } : undefined;
+    }
+
+    parseGetDFUWorkunitsResponse(prefix, response) {
+        return response?.GetDFUWorkunitsResponse?.results?.DFUWorkunit.map(item => {
+            return {
+                storeID: ++this._rowID,
+                id: item.ID,
+                Type: nlsHPCC.DFUWorkunit,
+                Reason: prefix,
+                Summary: item.ID,
+                _type: "DFUWuid",
+                _wuid: item.ID,
+                ...item
+            };
+        });
+    }
+
+    parseDFUQueryResponse(prefix, response) {
+        return response?.DFUQueryResponse?.DFULogicalFiles?.DFULogicalFile.map(item => {
+            if (item.isSuperfile) {
+                return {
                     storeID: ++this._rowID,
-                    id: this.id + item.Wuid,
-                    Type: nlsHPCC.ECLWorkunit,
+                    id: item.Name,
+                    Type: nlsHPCC.SuperFile,
                     Reason: prefix,
-                    Summary: item.Wuid,
-                    _type: "Wuid",
-                    _wuid: item.Wuid
-                });
-                this.eclStore.add(ESPWorkunit.Get(item.Wuid, item), { overwrite: true });
+                    Summary: item.Name,
+                    _type: "SuperFile",
+                    _nodeGroup: item.NodeGroup,
+                    _name: item.Name,
+                    ...item
+                };
+            }
+            return {
+                storeID: ++this._rowID,
+                id: item.NodeGroup + "::" + item.Name,
+                Type: nlsHPCC.LogicalFile,
+                Reason: prefix,
+                Summary: item.Name + " (" + item.NodeGroup + ")",
+                _type: "LogicalFile",
+                _nodeGroup: item.NodeGroup,
+                _name: item.Name,
+                ...item
+            };
+        });
+    }
+
+    parseWUListQueriesResponse(prefix, response) {
+        return response?.WUListQueriesResponse?.QuerysetQueries?.QuerySetQuery.map(item => {
+            return {
+                storeID: ++this._rowID,
+                id: item.QuerySetId + "::" + item.Id,
+                Type: nlsHPCC.Query,
+                Reason: prefix,
+                Summary: item.Name + " (" + item.QuerySetId + " - " + item.Id + ")",
+                _type: "Query",
+                _querySetId: item.QuerySetId,
+                _id: item.Id,
+                ...item
+            };
+        });
+    }
+
+    loadWUQueryResponse(results) {
+        if (results) {
+            results.forEach((item, idx) => {
+                this.store.put(item, { overwrite: true });
+                this.eclStore.put(ESPWorkunit.Get(item._wuid, item), { overwrite: true });
             });
-            return workunits.length;
+            return results.length;
         }
         return 0;
     }
 
-    loadGetDFUWorkunitResponse(prefix, response) {
-        const workunit = response?.GetDFUWorkunitResponse?.result;
-        if (workunit && workunit.State !== 999) {
-            this.store.add({
-                storeID: ++this._rowID,
-                id: this.id + workunit.ID,
-                Type: nlsHPCC.DFUWorkunit,
-                Reason: prefix,
-                Summary: workunit.ID,
-                _type: "DFUWuid",
-                _wuid: workunit.ID
-            });
-            this.dfuStore.add(ESPDFUWorkunit.Get(workunit.ID, workunit), { overwrite: true });
+    loadGetDFUWorkunitResponse(workunit) {
+        if (workunit) {
+            this.store.put(workunit, { overwrite: true });
+            this.dfuStore.put(ESPDFUWorkunit.Get(workunit._wuid, workunit), { overwrite: true });
             return 1;
         }
         return 0;
     }
 
-    loadGetDFUWorkunitsResponse(prefix, response) {
-        const workunits = response?.GetDFUWorkunitsResponse?.results?.DFUWorkunit;
-        if (workunits) {
-            workunits.forEach((item, idx) => {
-                this.store.add({
-                    storeID: ++this._rowID,
-                    id: this.id + item.ID,
-                    Type: nlsHPCC.DFUWorkunit,
-                    Reason: prefix,
-                    Summary: item.ID,
-                    _type: "DFUWuid",
-                    _wuid: item.ID
-                });
-                this.dfuStore.add(ESPDFUWorkunit.Get(item.ID, item), { overwrite: true });
+    loadGetDFUWorkunitsResponse(results) {
+        if (results) {
+            results.forEach((item, idx) => {
+                this.store.put(item, { overwrite: true });
+                this.dfuStore.put(ESPDFUWorkunit.Get(item._wuid, item), { overwrite: true });
             });
-            return workunits.length;
+            return results.length;
         }
         return 0;
     }
 
-    loadDFUQueryResponse(prefix, response) {
-        const items = response?.DFUQueryResponse?.DFULogicalFiles?.DFULogicalFile;
+    loadDFUQueryResponse(items) {
         if (items) {
             items.forEach((item, idx) => {
-                if (item.isSuperfile) {
-                    this.store.add({
-                        storeID: ++this._rowID,
-                        id: this.id + item.Name,
-                        Type: nlsHPCC.SuperFile,
-                        Reason: prefix,
-                        Summary: item.Name,
-                        _type: "SuperFile",
-                        _name: item.Name
-                    });
-                } else {
-                    this.store.add({
-                        storeID: ++this._rowID,
-                        id: this.id + item.Name,
-                        Type: nlsHPCC.LogicalFile,
-                        Reason: prefix,
-                        Summary: item.Name + " (" + item.NodeGroup + ")",
-                        _type: "LogicalFile",
-                        _nodeGroup: item.NodeGroup,
-                        _name: item.Name
-                    });
-                }
-                this.fileStore.add(ESPLogicalFile.Get(item.NodeGroup, item.Name, item), { overwrite: true });
+                this.store.put(item, { overwrite: true });
+                this.fileStore.put(ESPLogicalFile.Get(item._nodeGroup, item._name, item), { overwrite: true });
             });
             return items.length;
         }
         return 0;
     }
 
-    loadWUListQueriesResponse(prefix, response) {
-        const items = response?.WUListQueriesResponse?.QuerysetQueries?.QuerySetQuery;
+    loadWUListQueriesResponse(items) {
         if (items) {
             items.forEach((item, idx) => {
-                this.store.add({
-                    storeID: ++this._rowID,
-                    id: this.id + item.QuerySetId + "::" + item.Id,
-                    Type: nlsHPCC.Query,
-                    Reason: prefix,
-                    Summary: item.Name + " (" + item.QuerySetId + " - " + item.Id + ")",
-                    _type: "Query",
-                    _querySetId: item.QuerySetId,
-                    _id: item.Id
-                });
-                this.queryStore.add(ESPQuery.Get(item.QuerySetId, item.Id, item), { overwrite: true });
+                this.store.put(item, { overwrite: true });
+                this.queryStore.put(ESPQuery.Get(item._querySetId, item._id, item), { overwrite: true });
             });
             return items.length;
         }
@@ -273,86 +297,97 @@ export class ESPSearch {
 }
 
 export function searchAll(searchText: string,
-    loadWUQueryResponse: (what: string, response: any) => void,
-    loadGetDFUWorkunitResponse: (what: string, response: any) => void,
-    loadGetDFUWorkunitsResponse: (what: string, response: any) => void,
-    loadDFUQueryResponse: (what: string, response: any) => void,
-    loadWUListQueriesResponse: (what: string, response: any) => void,
+    loadWUQueryResponse: (prefix: string, response: any) => void,
+    loadGetDFUWorkunitResponse: (prefix: string, response: any) => void,
+    loadGetDFUWorkunitsResponse: (prefix: string, response: any) => void,
+    loadDFUQueryResponse: (prefix: string, response: any) => void,
+    loadWUListQueriesResponse: (prefix: string, response: any) => void,
     start: (searchCount: number) => void,
     done: (success: boolean) => void,
-): searchAllResponse {
+): any {
 
-    let searchECL = false;
-    let searchECLText = false;
-    let searchDFU = false;
-    let searchFile = false;
-    let searchQuery = false;
+    const generateSearchParams = (searchText: string): SearchParams => {
+        const searchParams: SearchParams = {
+            searchECL: false,
+            searchECLText: false,
+            searchDFU: false,
+            searchFile: false,
+            searchQuery: false,
+            text: searchText,
+            searchType: ""
+        };
 
-    let retVal: searchAllResponse = undefined;
+        if (searchText.indexOf("ECL:") === 0) {
+            searchParams.searchType = "ecl";
+            searchParams.searchECL = true;
+            searchParams.searchECLText = true;
+            searchParams.text = searchText.substring(4);
+        } else if (searchText.indexOf("DFU:") === 0) {
+            searchParams.searchType = "dfu";
+            searchParams.searchDFU = true;
+            searchParams.text = searchText.substring(4);
+        } else if (searchText.indexOf("FILE:") === 0) {
+            searchParams.searchType = "file";
+            searchParams.searchFile = true;
+            searchParams.text = searchText.substring(5);
+        } else if (searchText.indexOf("QUERY:") === 0) {
+            searchParams.searchType = "query";
+            searchParams.searchQuery = true;
+            searchParams.text = searchText.substring(6);
+        } else {
+            searchParams.searchECL = true;
+            searchParams.searchDFU = true;
+            searchParams.searchFile = true;
+            searchParams.searchQuery = true;
+        }
+        searchParams.text = searchParams.text.trim();
 
-    if (searchText.indexOf("ecl:") === 0) {
-        retVal = "ecl";
-        searchECL = true;
-        searchECLText = true;
-        searchText = searchText.substring(4);
-    } else if (searchText.indexOf("dfu:") === 0) {
-        searchDFU = true;
-        searchText = searchText.substring(4);
-    } else if (searchText.indexOf("file:") === 0) {
-        searchFile = true;
-        searchText = searchText.substring(5);
-    } else if (searchText.indexOf("query:") === 0) {
-        searchQuery = true;
-        searchText = searchText.substring(6);
-    } else {
-        searchECL = true;
-        searchDFU = true;
-        searchFile = true;
-        searchQuery = true;
+        return searchParams;
     }
-    searchText = searchText.trim();
 
     const searchArray = [];
-    if (searchECL) {
-        searchArray.push(WsWorkunits.WUQuery({ request: { Wuid: "*" + searchText + "*" }, suppressExceptionToaster: true }).then(function (response) {
+    const searchParams = generateSearchParams(searchText);
+
+    if (searchParams.searchECL) {
+        searchArray.push(WsWorkunits.WUQuery({ request: { Wuid: "*" + searchParams.text + "*" } }).then(response => {
             loadWUQueryResponse(nlsHPCC.WUID, response);
         }));
-        searchArray.push(WsWorkunits.WUQuery({ request: { Jobname: "*" + searchText + "*" } }).then(function (response) {
+        searchArray.push(WsWorkunits.WUQuery({ request: { Jobname: "*" + searchText + "*" } }).then(response => {
             loadWUQueryResponse(nlsHPCC.JobName, response);
         }));
-        searchArray.push(WsWorkunits.WUQuery({ request: { Owner: searchText } }).then(function (response) {
+        searchArray.push(WsWorkunits.WUQuery({ request: { Owner: searchText } }).then(response => {
             loadWUQueryResponse(nlsHPCC.Owner, response);
         }));
     }
-    if (searchECLText) {
-        searchArray.push(WsWorkunits.WUQuery({ request: { ECL: searchText } }).then(function (response) {
+    if (searchParams.searchECLText) {
+        searchArray.push(WsWorkunits.WUQuery({ request: { ECL: searchParams.text } }).then(response => {
             loadWUQueryResponse(nlsHPCC.ECL, response);
         }));
     }
-    if (searchDFU) {
-        searchArray.push(FileSpray.GetDFUWorkunit({ request: { wuid: "*" + searchText + "*" }, suppressExceptionToaster: true }).then(function (response) {
-            loadGetDFUWorkunitResponse(nlsHPCC.WUID, response);
+    if (searchParams.searchDFU) {
+        searchArray.push(FileSpray.GetDFUWorkunit({ request: { wuid: "*" + searchText + "*" } }).then(response => {
+            loadGetDFUWorkunitResponse(nlsHPCC.ECL, response);
         }));
-        searchArray.push(FileSpray.GetDFUWorkunits({ request: { Jobname: "*" + searchText + "*" } }).then(function (response) {
+        searchArray.push(FileSpray.GetDFUWorkunits({ request: { Jobname: "*" + searchText + "*" } }).then(response => {
             loadGetDFUWorkunitsResponse(nlsHPCC.JobName, response);
         }));
-        searchArray.push(FileSpray.GetDFUWorkunits({ request: { Owner: searchText } }).then(function (response) {
+        searchArray.push(FileSpray.GetDFUWorkunits({ request: { Owner: searchText } }).then(response => {
             loadGetDFUWorkunitsResponse(nlsHPCC.Owner, response);
         }));
     }
-    if (searchFile) {
-        searchArray.push(WsDfu.DFUQuery({ request: { LogicalName: "*" + searchText + "*" } }).then(function (response) {
+    if (searchParams.searchFile) {
+        searchArray.push(WsDfu.DFUQuery({ request: { LogicalName: "*" + searchParams.text + "*" } }).then(response => {
             loadDFUQueryResponse(nlsHPCC.LogicalName, response);
         }));
-        searchArray.push(WsDfu.DFUQuery({ request: { Owner: searchText } }).then(function (response) {
+        searchArray.push(WsDfu.DFUQuery({ request: { Owner: searchParams.text } }).then(response => {
             loadDFUQueryResponse(nlsHPCC.Owner, response);
         }));
     }
-    if (searchQuery) {
-        searchArray.push(WsWorkunits.WUListQueries({ request: { QueryID: "*" + searchText + "*" } }).then(function (response) {
+    if (searchParams.searchQuery) {
+        searchArray.push(WsWorkunits.WUListQueries({ request: { QueryID: "*" + searchParams.text + "*" } }).then(response => {
             loadWUListQueriesResponse(nlsHPCC.ID, response);
         }));
-        searchArray.push(WsWorkunits.WUListQueries({ request: { QueryName: "*" + searchText + "*" } }).then(function (response) {
+        searchArray.push(WsWorkunits.WUListQueries({ request: { QueryName: "*" + searchParams.text + "*" } }).then(response => {
             loadWUListQueriesResponse(nlsHPCC.Name, response);
         }));
     }
@@ -364,5 +399,5 @@ export function searchAll(searchText: string,
         done(false);
     });
 
-    return retVal;
+    return searchParams.searchType;
 }

